src/Validator.php

```diff
@@ -105,20 +85,28 @@ public function optional($key, $name = null, $allowEmpty = true)
     public function validate(array $values, $context = self::DEFAULT_CONTEXT)
     {
         $isValid = true;
-        $messageStack = $this->buildMessageStack($context);
-        $this->output = new Container();
+        $output = new Container();
         $input = new Container($values);
+        $stack = $this->getMessageStack($context);
+
+        if ($context !== self::DEFAULT_CONTEXT) {
+            $stack->merge($this->getMessageStack(self::DEFAULT_CONTEXT));
+        }
 
         foreach ($this->chains[$context] as $chain) {
             /** @var Chain $chain */
-            $isValid = $chain->validate($messageStack, $input, $this->output) && $isValid;
+            $isValid = $chain->validate($stack, $input, $output) && $isValid;
         }
 
-        return new ValidationResult(
+        $result = new ValidationResult(
             $isValid,
-            $this->messageStack->getMessages(),
-            $this->output->getArrayCopy()
+            $stack->getMessages(),
+            $output->getArrayCopy()
         );
+
+        $stack->reset();
```

Comment:
Well, the stack is being built at the moment that you start validating. If you were to call `$validator->validate()` twice, it will re-use the same MessageStack object, meaning the messages of the first validation run would appear in the second validation run -- that's not what you want. What you *do* want is to keep the message overwrites, but not the messages themselves. That's why ;)